accepted/0000-install-peer-deps.md

```diff
@@ -0,0 +1,144 @@
+`peerDependencies` by themselves, prompted by a warning.
+
+That warning is often misinterpreted as a problem, and reported to package
+maintainers, who in response, sometimes omit the peer dependency, treating
+it as effectively an optional dependency instead, but with no checks on its
+version range or validity.
+
+Furthermore, since the npm installer is not peer dependency aware, it
+can design a tree which causes problems when peer dependencies are present.
+
+This proposed algorithm addresses these problems, making `peerDependencies`
+a first-class concept and a requirement for package tree validity.
+
+For example, `tap` had a dependency on `ink`, which had a peer dependency on
+`react@16`.  In order to meet this peer dependency `tap` also added a
+dependency on `react@16`.  However, if a package depends on both `tap` and
+`react@15`, then the installer will see the conflicts _only as it relates
+to tap's dependency_, resulting in a package tree like:
+
+```
++-- react (15)
++-- ink
++-- tap
+    +-- react (16)
+```
+
+Because no version of `ink` existed higher in the tree, the installer
+moves it up a level, even though this breaks the peer dependency.
+
+To work around this, `tap` currently bundles both `ink` and `react`, but
+this is not optimal.  In cases where `ink` and/or `react` _can_ be
+deduplicated, they no longer are.
+
+## Detailed Explanation
+
+This extends the "maximally naive deduplication" algorithm that npm
+currently uses.
+
+### Validity Test
+
+A peer dependency is valid iff:
+
+- The name resolves from the dependant package to a package which satisfies
+  the listed dependency according to standard dependency resolution
+  semanatics, and
+- The resolved dependency is not found in the dependant's `node_modules`
+  tree (ie, it must be at or above it's own parent), _unless_ the dependent
+  is the root in its package tree.
+
+### Adding a New Dep
+
+When adding a dependency `D` in a range `R` with a set of peer dependencies
+`P` at location `L` in the tree:
+
+- For each `p` in `P`, starting from `L`, find the location in the
+  tree closest to the root where `p` can be placed without conflicts.
+- If all `p` in `P` can be placed:
+    - then: note the location furthest from the root where some `p` was
+      placed, as location `L'`
+    - else: error, `D` cannot be placed in this tree at location `L`.
+- Starting from `L`, find the location in the tree closest to `L'` where
+  `D` can be placed without conflicts.
+- If `D` can be placed between `L` and `L'`:
+  - then: hooray!  it is installed successfully.
+  - else: error, `D` cannot be placed in this tree at location `L`.
+
+(Optional failure handling: attempt with other versions of `D` in the range
+`R`.)
+
+### Handling Future Tree Munging
+
+If a user installs a new dependency, which will cause a conflict with
+`D` or any of `P`, then re-start the placement of `D` and `P` at `L`.
+
+If `D` and `P` cannot be placed in the tree in the presence of the newly
```

Comment:
> If `D` and `P` cannot be placed in the tree in the presence of the newly requested dependency, then refuse to install it until the user resolves the conflict

How would the user know there is a conflict if this is the case? Would it throw some error that is shown in the console?

Reply:
Yeah, the CLI would say "hey, buddy, you can't install that there, because it conflicts with XYZ".

To use the example elsewhere in the rfc, if you install `ink`, then try to install `react@15`, it'll say "no, that conflicts with ink's peer dependency on react@16.  Remove it or try again."

If you depended on `tap` which depends on `ink`, then it'd move `ink` and `react@16` underneath tap.